app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,151 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.home.blocklist
+
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.home.HomeFragmentAction
+import org.mozilla.fenix.home.HomeFragmentState
+import org.mozilla.fenix.home.recentbookmarks.RecentBookmark
+import org.mozilla.fenix.home.recenttabs.RecentTab
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem
+import org.mozilla.fenix.utils.Settings
+
+/**
+ * This [Middleware] reacts to item removals from the home screen, adding them to a blocklist.
+ * Additionally, it reacts to state changes in order to filter them by the blocklist.
+ *
+ * @param settings Blocklist is stored here as a string set
+ */
+class BlocklistMiddleware(
+    private val settings: Settings
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    private var blocklist: Set<String> = settings.homescreenBlocklist
+
+    /**
+     * Will filter "Change" actions using the blocklist and use "Remove" actions to update
+     * the blocklist.
+     */
+    override fun invoke(
+        context: MiddlewareContext<HomeFragmentState, HomeFragmentAction>,
+        next: (HomeFragmentAction) -> Unit,
+        action: HomeFragmentAction
+    ) {
+        val state = context.state
+        next(
+            when (action) {
+                is HomeFragmentAction.Change -> {
+                    action.copy(
+                        recentBookmarks = action.recentBookmarks.filteredByBlocklist(blocklist),
+                        recentTabs = action.recentTabs.filteredByBlocklist(blocklist),
+                    )
+                }
+                is HomeFragmentAction.RecentTabsChange -> {
+                    action.copy(
+                        recentTabs = action.recentTabs.filteredByBlocklist(blocklist)
+                    )
+                }
+                is HomeFragmentAction.RecentBookmarksChange -> {
+                    action.copy(
+                        recentBookmarks = action.recentBookmarks.filteredByBlocklist(blocklist)
+                    )
+                }
+                is HomeFragmentAction.RecentHistoryChange -> {
+                    action.copy(recentHistory = action.recentHistory.filteredByBlocklist(blocklist))
+                }
+                is HomeFragmentAction.RemoveRecentTab -> {
+                    if (action.recentTab is RecentTab.Tab) {
+                        val url = action.recentTab.state.content.url
+                        val updatedBlocklist = blocklist + url
+                        settings.homescreenBlocklist = updatedBlocklist
```

Comment:
I recalled @csadilek saying that desktop encrypts the sites that are blocked by the user. I imagine we should do the same and not store the set as plaintext.

Reply:
It's noted in the [Jira ticket](https://mozilla-hub.atlassian.net/browse/FNXV2-18720) that desktop stores the list as a comma-separated list of hashes, but I didn't think it was a necessarily a requirement. I assumed it was done that way to speed up comparisons, not necessarily for encryption. I figured the size of the list was likely to remain relatively small and I wasn't sure how to solve the hash collision problem, so I went with the straightforward implementation

Reply:
Let's check with @csadilek to make sure. I can imagine we might not want to store a list of plaintext user blocked website. Often, following the desktop implementation is an implicit requirement. It's just that not everyone is aware of the desktop details and can often be missed in the product requirements.